container_api/src/interface.rs

```diff
@@ -18,6 +22,35 @@ pub trait DispatchRpc {
     fn handler(self) -> IoHandler;
 }
 
+macro_rules! container_call {
+    ( |$container:ident| $call_expr:expr ) => {
+        match * CONTAINER.lock().unwrap() {
+            Some( ref mut $container) => {
+                $call_expr
+                    .map_err( | e | {
+                        let mut new = jsonrpc_core::Error::internal_error();
+                        new.message = e.to_string();
+                        new
+                    })
+            }
+            None => {
+                println!("Admin container function called without a container mounted as singleton!");
+                // If interfaces are supposed to work, the container needs to be mounted to a static place
+                // with container_api::container::mount_container_from_config(config: Configuration).
+                // There are cases in which we don't want to treat the container as a singleton such as
+                // holochain_nodejs and tests in particular. In those cases, calling admin functions via
+                // interfaces (websockt/http) won't work, but also we don't need that.
```

Comment:
The named mock network thing was a total hack, because MockSingleton was already firmly embedded in the architecture and would have taken too much work to dislodge it. I would have preferred that entities that need a mock network would own their own reference to the network they're on, rather than referring to a string. The container<->interface relationship is still new enough that it would be possible to represent the relationship with proper references, and I hope that if and when we want to have multiple containers, that's still the case, and we don't have to resort to magic strings.

> From perspective of the interface (thread) there should clearly be just one container. Hence singleton.

Agree with the first sentence but not the second. Yes, from the perspective of any interface there should be one container. Interfaces and Containers are many-to-one. In Rust types that means the interface should own a single reference to a container, nothing more. A singleton is sufficient but not necessary to represent a many-to-one relationship. It just says that every Interface refers to the one singleton Container, rather than one arbitrary container.